### PR TITLE
Add :service_prefix option for service names

### PR DESCRIPTION
**Pull request by delner**
Following up on the request from #513 , this pull request adds an option to the configuration that allows you to define a prefix to add to all service names by default.

```ruby
Datadog.configure do |c|
  c.tracer service_prefix: 'myapp_'
  c.use :sinatra
end
```

The above sample will produce Sinatra spans with `myapp_sinatra`.

### CODE REVIEW

File: lib/ddtrace/contrib/configuration/settings.rb [hunk 1]
```diff
@@ -29,6 +29,16 @@ def [](name)
         def []=(name, value)
           respond_to?("#{name}=") ? send("#{name}=", value) : set_option(name, value)
         end
+
+        # Override set_option to allow service_prefix to be added.
+        def set_option(name, value)
+          if name.to_sym == :service_name && !value.nil?
+            tracer = get_option(:tracer)
+            super(name, "#{tracer.service_prefix}#{value}")
```
Comment (delner) [inline lib/ddtrace/contrib/configuration/settings.rb]: `service_prefix` lives on the Tracer, despite not actually being used at all by the tracer. Then this configuration object is the only thing to actually reference and use it, creating a dependency that might not make sense. We're really just putting this on `Tracer` because `c.tracer` in the configuration API acts as a "general settings" right now: whatever isn't integration specific ends up there.

We might want to consider moving `service_prefix` to some kind of new "general settings" object, which could contain settings not actually applicable to the tracer or integrations themselves. It would help us avoid unnecessary entanglement between the tracer and the configuration parts of the code.

**Comment by sponomarev** [comment]
I hope this PR will be merged sometime

**Comment by frsantos** [comment]
With the recent changes in configuration, is there any possibility of continuing this PR?

**Comment by delner** [comment]
@frsantos I think this might soon become obsolete.

We introduced `c.service` in 0.34.0 which allows you to define a default service name for the application. Currently, this service name is overridden by most integrations with their own default service names, but we're going to change this in the near future.

Currently what we're thinking is:

 - "Internal" integrations (that generate spans that happen entirely within the Ruby application) would no longer define `service_name`: they would use `c.service_name`.
 - "External" integrations (that generate spans that wrap calls to external processes e.g. databases, HTTP, etc) might continue to define their own service name and these spans would get both associated with the default service name and the one defined on the integration. The idea would be that you could see the span was created by `billing-api` but actually instrumented a call to the `accounts-db`, or something like that. We haven't yet committed to this kind of behavior, so details/implementation is pending here.

These changes (or whatever we end up with) would likely cause breaking changes for the Ruby tracer as we redefine service name behavior, and because we want to minimize these kind of changes, we would want to group any similar behavior in the same release (e.g. service name prefixes, if still relevant.)

TL;DR Basically we're working on larger changes that might encompass this or make it obsolete, so we probably still want to wait on it for now.

**Comment by MXfive** [comment]
Hey @delner We already have Datadog working great in production with a single service name per integration as per the defaults of this gem.  This is quite confusing when trying to on board our team and working with other datadog integrations though.  So we are currently trying to mimic this setup of internal and external services by manually naming all the "internal" integrations with the same name and the "external ones" with appropriate different names.  However when we actually try to run our app like this in staging, the metrics for all of the "non-primary" operations vary between completely missing and very incomplete.

We have been trying to work with your support to get to the bottom of this, but not getting anywhere so far.  Is there anything else that needs to be done on the Ruby integration side to make the non-primary operations behave correctly once they reach Datadog?

**Comment by delner** [comment]
When this PR was opened, we intended to split a single Ruby application into many "services", both "internal" and "external", in order to provide layers of visibility on key metrics (e.g. Rack vs GraphQL vs Rails vs ActiveRecord and so on...)

This approach has two major downsides; 1) it requires users to label each of their integrations with service names individually and to differentiate "internal" vs "external" integrations which complicates onboarding/setup, and 2) can create a mess of the service map in the Datadog UI... where users typically expect to see `service = application`, they were seeing potentially dozens of "services" instead.

Since then, we've decided it would be better if each Ruby application had only one service name. Ideally in the near future, all spans produced by a Ruby app will receive this single service name via `DD_SERVICE` or configuration file. This pull request doesn't really jive with that direction, so I've decided to close this for now.

If you still have other thoughts, please feel free to share them here. We'd like to make "service name" intuitive, fluid, and work for your application, and user feedback helps us do just that.

**Comment by MXfive** [comment]
@delner There is functionality that due to Datadog only recording metrics from trace's primary operations (`trace.#{operation_name}.hits`, `trace.#{operation_name}.duration`, `trace.#{operation_name}.errors`, etc.) tagged by resources.

We get around this at no additional cost by naming each integration within an application differently as described in this ticket.

We rely on this heavily for things like dashboarding and alerting on throughput, latency and error rate of the RPC client in a specific application calling a specific remote resource.

So I can only really agree on this direction if there are also plans to start generating metrics on non primary operations on Datadog's end.

**Comment by marcotc** [comment]
@MXfive we have the ability to generate metrics on non-root spans by marking them as "measured".
This feature was introduced version [0.34.0](https://github.com/DataDog/dd-trace-rb/releases/tag/v0.34.0) (by https://github.com/DataDog/dd-trace-rb/pull/974).

That work "marked" all spans that we considered at the time important with the "measured" tag. You can see all places where we mark them here: https://github.com/DataDog/dd-trace-rb/search?q=set_measured+path%3Alib%2Fddtrace%2Fcontrib

If you already using 0.34.0 or newer and are not seeing metrics for spans that you'd like to alert on, please let us know by opening a GitHub issue so we can address it. It is definitely our goal to have every operation that one would reasonably monitor to generate rich metrics.

**Comment by MXfive** [comment]
@marcotc That is exactly what I was looking for, thanks a lot for your reply.  I recall now when doing the initial integration with dd-trace-rb being confused as to why I seemed to be getting some metrics for only some of the non primary spans (the template rendering), but not on the ones I wanted.  Because it was taking some time already because of some other bugs to get integrated, I went with the separate service names as this functionality is not documented.

I do think enabling this for RPC integrations is a sensible default as there is a lot of value that can be had from those metrics for monitoring and alerting.

**Comment by marcotc** [comment]
> I do think enabling this for RPC integrations is a sensible default as there is a lot of value that can be had from those metrics for monitoring and alerting.

@MXfive I agree that RPC integrations require this kind of visibility. The only true RPC integration we currently have is gRPC, and [it marks its span as measured](https://github.com/DataDog/dd-trace-rb/blob/e63e65f19887d011cd957f7e58361f4af0df2050/lib/ddtrace/contrib/grpc/datadog_interceptor/server.rb#L58). 
Were you referring to a different integration point, or different gem?

**Comment by MXfive** [comment]
@marcotc That is measuring only the server component of the gRPC gem, I was referring to[ the client](https://github.com/DataDog/dd-trace-rb/blob/e63e65f19887d011cd957f7e58361f4af0df2050/lib/ddtrace/contrib/grpc/datadog_interceptor/client.rb) and other HTTP clients (Exxon, Faraday, rest_client, etc.)

**Comment by marcotc** [comment]
This makes sense @MXfive.
When we implemented `measured`, the idea was that spans that represented access to external services would have their own service name, thus naturally having metrics calculated for that separate service name. Having both `measured` and separate service name metrics at the same time can cause double metric counting, due to name collision. So we chose to not tag any requests that perform external work with `measured`.

Recently, there has been discussion about changing this approach, as reports similar to your have started surfacing more and we are working on solving the overlapping metric issue. So, today, we have internal work ongoing to address it.

But until we are able to safely make that change, using a separate service name for any service that you require metrics for, and don't see them today, is our recommendation.

**Comment by agrobbin** [comment]
@delner @marcotc for those who want to move to this newer approach of consolidating services in APM, what would you recommend? We currently have an application that uses Rails (duh 😄 ), Sidekiq, AWS, Redis, and a few other integrations for external HTTP client tracing, in addition to 1 other process type for [sidekiq_publisher](https://github.com/ezcater/sidekiq_publisher) (which has its own Datadog instrumentation), and currently the service map is, as you said above, *a mess*. I saw that `DD_SERVICE` was introduced in #983, but with that being over a year ago, I'm wondering what the recommendation is for unifying services like this:

* `aws`
* `active_record`
* `sidekiq` (server)
* `sidekiq-client`
* `[app]`
* `[app]-action_cable`
* `sidekiq-publisher`

This does not include Redis, Postgres, nor a bunch of services for external systems, which make a lot of sense to me to continue to be split and maintained as separate services in APM. However, the service naming approach gets even more complicated when we add another application to APM, which we plan on doing very soon.

The naming inconsistencies, in addition to the sheer number of different services for the same application, would be great to simplify, and if there are recommendations for how best to configure dd-trace, that would be great!
